tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreStore.java

```diff
@@ -0,0 +1,231 @@
+    QueryResults<Key> results = datastore.run(query);
+    keys = Streams.stream(results)
+        .map(key -> key.getNameOrId().toString())
+        .toArray(String[]::new);
+
+    if (keys == null) {
+      keys = new String[0];
+    }
+
+    return keys;
+  }
+
+  /**
+   * Load and return the Session associated with the specified session identifier from this Store,
+   * without removing it. If there is no such stored Session, return null.
+   *
+   * <p>Look in the datastore for a serialized session and attempt to deserialize it.</p>
+   *
+   * <p>If the session is successfully deserialized, it is added to the current manager and is
+   * returned by this method. Otherwise null is returned </p>
+   *
+   * @param id Session identifier of the session to load
+   * @return The loaded session instance
+   * @throws ClassNotFoundException If a deserialization error occurs
+   */
+  @Override
+  public Session load(String id) throws ClassNotFoundException, IOException {
+    log.debug("Session " + id + " requested");
+    StandardSession session = null;
+
+    Entity sessionEntity = datastore.get(keyFactory.newKey(id));
+
+    if (sessionEntity != null) {
+      try {
+        InputStream fis = sessionEntity.getBlob("content").asInputStream();
+        ObjectInputStream ois = getObjectInputStream(fis);
+        session = (StandardSession) manager.createEmptySession();
+        session.readObjectData(ois);
+        session.setManager(manager);
+      } catch (IOException e) {
+        log.warn("An error occurred during session deserialization");
+        session = null;
+      }
+    }
+
+    log.debug("Session " + id + " loaded");
+    return session;
+  }
+
+  /**
+   * Remove the Session with the specified session identifier from this Store, if present.
+   * If no such Session is present, this method takes no action.
+   *
+   * @param id Session identifier of the session to remove
+   */
+  @Override
+  public void remove(String id) {
+    log.debug("Removing session: " + id);
+    datastore.delete(keyFactory.newKey(id));
+  }
+
+  /**
+   * Remove all Sessions from this Store.
+   */
+  @Override
+  public void clear() throws IOException {
+    log.debug("Deleting all sessions");
+    datastore.delete(Arrays.stream(keys())
+                           .map(keyFactory::newKey)
+                           .toArray(Key[]::new));
+  }
+
+  /**
+   * Save the specified Session into this Store. Any previously saved information for
+   * the associated session identifier is replaced.
+   *
+   * <p>Attempt to serialize the session and send it to the datastore.</p>
+   *
+   * @throws IOException If an error occur during the serialization of the session.
+   *
+   * @param session Session to be saved
+   */
+  @Override
+  public void save(Session session) throws IOException {
+    log.debug("Persisting session: " + session.getId());
+
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
```

Comment:
Do you need to close these streams when you're done with them? I think this could be a potential memory leak.